impeller/renderer/backend/metal/surface_mtl.mm

```diff
@@ -248,13 +248,17 @@
   }
 
   if (drawable_) {
-    TRACE_EVENT0("flutter", "waitUntilScheduled");
     id<MTLCommandBuffer> command_buffer =
         ContextMTL::Cast(context.get())
             ->CreateMTLCommandBuffer("Present Waiter Command Buffer");
-    [command_buffer commit];
-    [command_buffer waitUntilScheduled];
-    [drawable_ present];
+    if ([[NSThread currentThread] isMainThread]) {
+      [command_buffer commit];
+      [command_buffer waitUntilScheduled];
+      [drawable_ present];
+    } else {
+      [command_buffer presentDrawable:drawable_];
```

Comment:
Oh wait, I need to double check if this messes up frame capture. If so we need to check if we're capturing a frame and block.

Reply:
Done